flang/lib/Lower/FIRBuilder.cpp

```diff
@@ -434,6 +441,51 @@ struct CharacterOpsBuilderImpl {
 
     return {substringRef, substringLen};
   }
+
+  mlir::Value createLenTrim(Char str) {
+    // Note: Runtime for LEN_TRIM should also be available at some
+    // point. For now use an inlined implementation.
+    auto indexType = builder.getIndexType();
+    mlir::Value len = builder.createHere<fir::ConvertOp>(indexType, str.len);
+    auto data = materializeValue(str);
+    auto one = builder.createIntegerConstant(indexType, 1);
+    auto i1Type = mlir::IntegerType::get(1, builder.getContext());
+    auto trueVal = builder.createIntegerConstant(i1Type, 1);
+    auto blank = createBlankConstant(str.getCharacterType());
+    // FIXME: Conversion wrong in general, it would be cool to compare fir.char
+    // without cast. Needs the kindmap.
+    auto i8Type = mlir::IntegerType::get(8, builder.getContext());
+    blank = builder.createHere<fir::ConvertOp>(i8Type, blank);
+    // initVal must not be random to account for the zero length case where the
+    // no iteration of the loop is ever executed.
+    mlir::Value initVal = builder.createHere<mlir::SubIOp>(len, one);
+
+    auto iterWhile =
+        builder.createHere<fir::IterWhileOp>(one, len, one, trueVal, initVal);
+    auto insPt = builder.saveInsertionPoint();
+    builder.setInsertionPointToStart(iterWhile.getBody());
+
+    auto index = iterWhile.getInductionVar();
+    auto charPosition = builder.createHere<mlir::SubIOp>(len, index);
+    auto c = createLoadCharAt(data, charPosition);
+    // FIXME: needs more generic conversion.
+    c = builder.createHere<fir::ConvertOp>(i8Type, c);
+    auto isBlank =
+        builder.createHere<mlir::CmpIOp>(mlir::CmpIPredicate::eq, blank, c);
+    llvm::SmallVector<mlir::Value, 2> results = {isBlank, charPosition};
```

Comment:
Is the `charPosition` really an induction? It appears that it is computed each iteration from `len` and the loop iteration value.

Reply:
This was changed following Val's comment, the loop is now decrementing and directly its induction is directly used (renamed `index`).

Reply:
Ignore my first , my test was too shallow, it appears `fir.iterate_while` cannot be used with a negative step, so I changed that, but my understanding of what was returned was incorrect, I thought it was not the induction variable but whatever index was passed in result.
Now, I ma actually getting `error: 'fir.result' op parent of result must have same arity`, which I suspect has to do with the way I am querying the iteration at the end (`iterWhile.getResult(1)`), I think I was using `iter_args` which is actually for additional arguments/results, hence, that worked before my changes to remove the extra result.
How do we get "%iterateResult" then (I only see `getInductionVar`, but I would not use that outside of the body) ?